src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+    }
+
+    private void getSampleRangesForValidationChecks(
+        Optional<Long> latestTime,
+        AnomalyDetector detector,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        Map<String, Object> topEntity
+    ) {
+        if (!latestTime.isPresent()) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        } else if (latestTime.get() <= 0) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        }
+        long timeRangeEnd = Math.min(Instant.now().toEpochMilli(), latestTime.get());
+        try {
+            getBucketAggregates(timeRangeEnd, listener, topEntity);
+        } catch (IOException e) {
+            listener.onFailure(new EndRunException(detector.getDetectorId(), CommonErrorMessages.INVALID_SEARCH_QUERY_MSG, e, true));
+        }
+    }
+
+    private void getBucketAggregates(
+        long latestTime,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        Map<String, Object> topEntity
+    ) throws IOException {
+        List<String> featureFields = ParseUtils.getFeatureFieldNames(anomalyDetector, xContentRegistry);
+        AggregationBuilder aggregation = getBucketAggregation(latestTime);
+        BoolQueryBuilder query = QueryBuilders.boolQuery().filter(anomalyDetector.getFilterQuery());
+        if (anomalyDetector.isMultientityDetector()) {
+            if (topEntity.isEmpty()) {
+                listener
+                    .onFailure(
+                        new ADValidationException(
+                            CommonErrorMessages.CATEGORY_FIELD_NO_DATA,
+                            DetectorValidationIssueType.CATEGORY,
+                            ValidationAspect.MODEL
+                        )
+                    );
+                return;
+            }
+            for (Map.Entry<String, Object> entry : topEntity.entrySet()) {
+                query.filter(QueryBuilders.termQuery(entry.getKey(), entry.getValue()));
+            }
+        }
+        if (anomalyDetector.isMultiCategoryDetector()) {
+            for (String category : anomalyDetector.getCategoryField()) {
+                query.filter(QueryBuilders.existsQuery(category));
+            }
+        }
+        for (String featureField : featureFields) {
+            query.filter(QueryBuilders.existsQuery(featureField));
+        }
```

Comment:
Same question for this filter

Reply:
This filter makes sure that the fields given in the feature queries are present in the documents that I am looking for in the bucket aggregation

Reply:
Ah, I see this isn't in the HCAD block. Still seems unnecessary in the HCAD scenario, can it be added as an `else` block so it's only added in the non-HCAD scenario?

Reply:
I still think its necessary in the HCAD scenario as the fields given in the feature query don't necessarily have to relate to the category fields given. For example the feature query could be aggregating `memory` field but up category field is `ip`. I still want to only look and see if we have enough documents with the `memory` field present in them when aggregating.